File: JavaScript/7js 3/reverseSlice.js
/*
|--------------------------------------------------------------------------
| String reverse slicing 101
|--------------------------------------------------------------------------
|
*/

function reverseSlice(str) {
   let res = [];
   for (let i = 0; i < str.length; i++) {
      res.push(
         str
            .split("")
            .splice(0, i + 1)
            .reverse()
            .join("")
      );
   }
   return res.reverse();
}


function reverseSlice(s) {
   return Array(s.length)
      .fill(0)
      .map((e, i) => s.split("").reverse().join("").slice(i));
}


function reverseSlice(str) {
   const res = [str[0]];
   for (let i = 1; i < str.length; i++) {
      res.unshift(str[i] + res[0]);
   }
   return res;
}


console.log(reverseSlice("123"), ["321", "21", "1"]);
console.log(reverseSlice("abcde"), ["edcba", "dcba", "cba", "ba", "a"]);

File: Python/7py 3/reverse_slice.py
"""
|--------------------------------------------------------------------------
| String reverse slicing 101
|--------------------------------------------------------------------------
|
"""

def reverse_slice(my_str):
   res = []
   for i in range(len(my_str)):
      res.append(''.join(reversed(my_str[:i+1])))
   return res[::-1]


def reverse_slice(s):
   return [s[::-1][i:] for i in range(len(s))]


def reverse_slice(s):
   a, i = [], len(s)
   while i>0:
      s=s[:i]
      a+=[s[::-1]]
      i-=1
   return a


print(reverse_slice('123'), ['321', '21', '1'])
print(reverse_slice('abcde'), ['edcba', 'dcba', 'cba', 'ba', 'a'])